File: BeerViewer/Models/Wrapper/Quest.cs
﻿using System;

using BeerViewer.Models.Enums;
using BeerViewer.Models.Raw;
using BeerViewer.Models.kcsapi;

namespace BeerViewer.Models.Wrapper
{
	public class Quest : SvData<kcsapi_quest>, IIdentifiable
	{
		public int Id => this.api_data.api_no;

		public QuestCategory Category => (QuestCategory)this.api_data.api_category;
		public QuestType Type => (QuestType)this.api_data.api_type;
		public QuestState State => (QuestState)this.api_data.api_state;
		public QuestProgress Progress => (QuestProgress)this.api_data.api_progress_flag;

		public string Title => this.api_data.api_title;
		public string Detail => this.api_data.api_detail.Replace("<br>", Environment.NewLine);

		public Quest(kcsapi_quest api_data) : base(api_data) { }

		public override string ToString()
			=> $"ID = {this.Id}, Category = {this.Category}, Title = \"{this.Title}\", Type = {this.Type}, State = {this.State}";
	}
}

File: WindowFrame/source/System/Homeport/Quest/Quest.ts
import { RawDataWrapper } from "System/Base/Wrapper";
import { IIdentifiable } from "System/Base/Interfaces/IIdentifiable";
import { QuestCategory, QuestProgress, QuestState, QuestType } from "System/Enums/Quests";
import { kcsapi_quest } from "System/Interfaces/kcsapi_quest";

class Quest extends RawDataWrapper<kcsapi_quest> implements IIdentifiable {
	public get Id(): number { return this.raw.api_no }

	public get Category(): QuestCategory { return this.raw.api_category }
	public get Type(): QuestType { return this.raw.api_type }
	public get State(): QuestState { return this.raw.api_state }
	public get Progress(): QuestProgress { return this.raw.api_progress_flag }

	public get Title(): string { return this.raw.api_title }
	public get Description(): string { return this.raw.api_detail.replace(/<br>/g, "\r\n") }

	constructor(api_data: kcsapi_quest) {
		super(api_data);
	}
}
export default Quest;